### PR TITLE
Too fast too free

### PR DESCRIPTION
**Pull request by isovector**
This PR is the first step towards sweet perf gains in `freer-simple`. I've left some commentary inline in the review on particularly prickly points that deserve discussion.

### CODE REVIEW

File: src/Control/Monad/Freer.hs [hunk 1]
```diff
@@ -217,165 +217,21 @@ module Control.Monad.Freer
   , run
   , runM
 
-    -- ** Building Effect Handlers
```
Comment (isovector) [inline src/Control/Monad/Freer.hs]: I chopped these out on the assumption that most people only want to use existing effects, not define new interpretations.

File: src/Control/Monad/Freer/Error.hs [hunk 2]
```diff
@@ -43,12 +44,12 @@ catchError
   => Eff effs a
   -> (e -> Eff effs a)
   -> Eff effs a
-catchError m handle = interposeWith (\(Error e) _ -> handle e) m
+catchError m handle = interceptRelay pure (\(Error e) _ -> handle e) m
```
Comment (isovector) [inline src/Control/Monad/Freer/Error.hs]: I haven't thought too much about this, but we can probably write it without `relay`. `relay` is *relay* slow.

File: src/Control/Monad/Freer/Internal.hs [hunk 3]
```diff
@@ -28,45 +28,31 @@
...
-  , replaceRelay
-  , replaceRelayS
-  , replaceRelayN
-
-    -- *** Low-level Functions for Building Effect Handlers
-  , qApp
-  , qComp
-
     -- ** Nondeterminism Effect
   , NonDet(..)
```
Comment (isovector) [inline src/Control/Monad/Freer/Internal.hs]: Why isn't this defined in `NonDet?`

File: src/Control/Monad/Freer/Internal.hs [hunk 4]
```diff
@@ -194,162 +184,27 @@ sendM = send
...
-  where
-    loop (Val x)  = pure' x
-    loop (E u' q) = case decomp u' of
-        Right x -> bind x k
-        Left  u -> E (weaken u) (tsingleton k)
-      where
-        k = qComp q loop
-{-# INLINE replaceRelay #-}
-
-replaceRelayN
```
Comment (isovector) [inline src/Control/Monad/Freer/Internal.hs]: I haven't reimplemented this in `too-fast-too-free` but it seems niche enough that we probably don't need to provide it?

File: src/Control/Monad/Freer/Interpretation.hs [hunk 5]
```diff
@@ -0,0 +1,213 @@
+{-# LANGUAGE DataKinds             #-}
+{-# LANGUAGE MonoLocalBinds        #-}
+{-# LANGUAGE QuantifiedConstraints #-}
+{-# LANGUAGE RankNTypes            #-}
+{-# LANGUAGE TypeOperators         #-}
+
+module Control.Monad.Freer.Interpretation where
```
Comment (isovector) [inline src/Control/Monad/Freer/Interpretation.hs]: This module has the interpreter combinators. I renamed them so their names are a little more meaningful to humans; I figured since the types have breaking changes anyway we might as well.

File: src/Control/Monad/Freer/NonDet.hs [hunk 6]
```diff
@@ -15,41 +15,33 @@ module Control.Monad.Freer.NonDet
...
-    loop jq (Val x) = pure (Just (x, msum jq))
-    loop jq (E u q) = case prj u of
-        Just MZero -> case jq of
-          []      -> pure Nothing
-          (j:jq') -> loop jq' j
-        Just MPlus -> loop (qApp q False : jq) (qApp q True)
-        Nothing    -> E u (tsingleton k)
-      where
-        k = qComp q (loop jq)
+msplit = undefined -- loop []
```
Comment (isovector) [inline src/Control/Monad/Freer/NonDet.hs]: Does anyone care about this? If so I'll figure out how to write it, but in the meantime cbf.

File: src/Control/Monad/Freer/State.hs [hunk 7]
```diff
@@ -94,14 +98,9 @@ transactionState
   -> Eff effs a
 transactionState m = do
     s0 <- get @s
-    (x, s) <- interposeS s0 (\s x -> pure (x, s)) handle m
+    (x, s) <- interceptS stateNat s0 m
```
Comment (isovector) [inline src/Control/Monad/Freer/State.hs]: I don't a million percent understand what `transactionState` is supposed to do, but presumably this does the same thing.

File: src/Control/Monad/Freer/Trace.hs [hunk 8]
```diff
@@ -28,7 +29,6 @@ trace :: Member Trace effs => String -> Eff effs ()
 trace = send . Trace
 
 -- | An 'IO' handler for 'Trace' effects.
-runTrace :: Eff '[Trace] a -> IO a
-runTrace (Val x) = return x
-runTrace (E u q) = case extract u of
-  Trace s -> putStrLn s >> runTrace (qApp q ())
+runTrace :: Member IO r => Eff (Trace ': r) ~> Eff r
+runTrace = natural @IO $ \(Trace s) -> putStrLn s
```
Comment (isovector) [inline src/Control/Monad/Freer/Trace.hs]: If this is going to be in the std lib, there really should be a `(w -> String) -> Writer w ~> Trace` interpreter

File: src/Control/Monad/Freer/Writer.hs [hunk 9]
```diff
@@ -33,5 +34,4 @@ tell w = send (Tell w)
 
 -- | Simple handler for 'Writer' effects.
 runWriter :: forall w effs a. Monoid w => Eff (Writer w ': effs) a -> Eff effs (a, w)
-runWriter = handleRelay (\a -> pure (a, mempty)) $ \(Tell w) k ->
-  second (w <>) <$> k ()
+runWriter = withStateful mempty $ \(Tell w) -> modify (<> w)
```
Comment (isovector) [inline src/Control/Monad/Freer/Writer.hs]: Does the original function do the monoid in the right direction? Seems like no?

File: src/Data/OpenUnion.hs [hunk 10]
```diff
@@ -23,6 +23,7 @@ module Data.OpenUnion
   , decomp
   , weaken
   , extract
+  , intro
```
Comment (isovector) [inline src/Data/OpenUnion.hs]: I'm strongly of the opinion that this entire module should be `Internal` --- real users have no use for `decomp`, `weaken`, `extract` or really anything other than `Member`.

File: stack.yaml [hunk 11]
```diff
@@ -1,4 +1,4 @@
-resolver: lts-12.18
+resolver: lts-13.0
```
Comment (isovector) [inline stack.yaml]: The `transform` combinator requires `-XQuantifiedConstraints`. We can specialize write `shortCircuit` and `stateful` directly in older versions of GHC to avoid this.

File: src/Control/Monad/Freer/Interpretation.hs [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+replace = naturally weaken
+{-# INLINE replace #-}
+
+
+------------------------------------------------------------------------------
+-- | Run an effect via the side-effects of a monad transformer.
+transform
+    :: ( MonadTrans t
+       , MFunctor t
+       , forall m. Monad m => Monad (t m)
```
Comment (isovector) [inline src/Control/Monad/Freer/Interpretation.hs]: This won't compile before GHC 8.6---we can just CPP it away for older versions.

File: src/Control/Monad/Freer/Interpretation.hs [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+    -> Eff r ~> Eff r
+intercept f (Freer m) = Freer $ \k -> m $ \u ->
+  case prj u of
+    Nothing -> k u
+    Just e  -> usingFreer k $ f e
+{-# INLINE intercept #-}
+
+------------------------------------------------------------------------------
+-- | Like 'interpret', but with access to intermediate state.
+interceptS
```
Comment (isovector) [inline src/Control/Monad/Freer/Interpretation.hs]: I'd like a better way of composing the interpreter combinators. It'd be cool to go `stateful . intercept` rather than needing to write `interceptS` and a bunch of other silly things.

File: src/Control/Monad/Freer/Internal.hs (hunk 4 above)
Comment (isovector) [reply src/Control/Monad/Freer/Internal.hs]: Nevermind. This is pretty useful! I'll reimplement it before merging.

File: src/Control/Monad/Freer/Interpretation.hs [hunk 14]
```diff
@@ -0,0 +1,211 @@
+{-# LANGUAGE CPP                   #-}
+{-# LANGUAGE DataKinds             #-}
+{-# LANGUAGE MonoLocalBinds        #-}
+#if __GLASGOW_HASKELL__ >= 806
+{-# LANGUAGE QuantifiedConstraints #-}
+#endif
+{-# LANGUAGE RankNTypes            #-}
+{-# LANGUAGE TypeOperators         #-}
```
Comment (lexi-lambda) [inline src/Control/Monad/Freer/Interpretation.hs]: A bunch of these are already on by default, so please remove the ones that aren’t strictly necessary.

File: src/Control/Monad/Freer/Interpretation.hs (hunk 5 above)
Comment (lexi-lambda) [reply src/Control/Monad/Freer/Interpretation.hs]: “Meaningful” is subjective. :) My experience is that some of the original names are more meaningful to a certain set of people, but perhaps less useful to others more familiar with other concepts. Here was my general philosophy for `freer-simple`’s names:

- A library ought to introduce as few concepts and as few terms of art as possible as long as they are entirely internally-consistent.
- When possible, terms should be borrowed from English words or phrases so as to borrow intuition via analogy.
- However, if an analogy is likely to break down quickly, especially if it is in confusing ways, then it’s more harmful than helpful. A fresh, artificial term should be used instead so as to avoid potentially-confusing baggage.

The metaphor I used for `freer-simple` is one that already exists in computer science: languages and interpreters. This metaphor is so ubiquitous that it appears in Gang of Four as the *interpreter pattern*, so it seems safe to assume any prior exposure is consistent with my use of the terms. In this analogy, effects are languages, and effect handlers are interpreters.

These are the only *fundamental* concepts in freer-simple’s vocabulary. Therefore, everything else really ought to fall out of application of the metaphor to the derived concepts. Let’s evaluate how well that’s worked, shall we?

- We’ll start with the simplest handler combinator, `interpret`. This name is essentially one of our axioms, so obviously it’s consistent with the metaphor.

- Next, let’s look at `translate`. I believe this to be a good name, since it fits into the metaphor very well. The process of moving from one language to another while preserving meaning is *translation*; that’s what this function does, so its name is good.

- What about `reinterpret` and friends? These names are a little less good upon close examination, but I think the metaphor still holds up. Morally, they are very much like `translate`, but since they run in a context in which all the other effects are accessible, they also have the ability to do some extra interpretation.

   In practice, `reinterpret` is far more useful than `translate`, since `translate` demands you use the effect constructors directly (e.g. `Ask` or `Get`), rather than using the wrapping functions (e.g. `ask` or `get`). Since effects are treated like abstractions, thinking about the constructors might be unnecessarily low-level. It’s a subtle distinction, one that I don’t expect most newcomers to immediately understand. Perhaps `translate` ought to be somehow deemphasized for that reason.

- `interpose` exists somewhat outside of the interpreter metaphor, instead living more inside the client/server metaphor that `handleRelay` uses. Perhaps a better name would be something like `embellish`, but honestly, that seems too cute to me, and the existing name still does make some sense: it’s interposing on an effect’s interpretation by some other interpreter.

- `subsume` needs little explanation, since it just does what it says. This name needs no metaphors!

- Finally, there are the “variant” handlers. `interpretM` interprets an effect monadically, so the name works. `interpretWith` and `interposeWith` are generalized versions of `interpret` and `interpose` that expose the continuation. The names don’t really connect to what they do (what does `With` have to do with being given the continuation?), so cleverer names would be better, but these are also the “advanced” handlers, which I don’t expect to be used by very many people, anyway.

All together, these names form what I think is a mostly-coherent linguistic abstraction. Ideally, a user of `freer-simple` should be thinking about an “effect” as a sort of primitive thing. In my ideal world, I would be able to make defining an effect use the syntax

```haskell
effect State s where
  get :: s
  put :: s -> ()
```

but I can’t, so I have to kludge it a bit. In any case, I’m really angling to make the abstraction as watertight as possible, without any strings attached.

---

Alright, so, what does that small essay have to do with your PR? Well, I want to give the context under which I’m evaluating your proposed changes. You propose a new function, `natural`, as well as a new name for `translate`, `replace`, but these don’t fit in with the metaphor at all. `natural` even demands that users be familiar with an unrelated concept, natural transformations. `replace` is vague and nonspecific (the name can come up *everywhere* in programming), and `natural` isn’t even a verb. The names aren’t *bad*, not in isolation, but they’re certainly inconsistent with the rest of the API.

I can make similar arguments for things like `relay`, `transform`, and `introduce`. The names aren’t *bad*, but they don’t obey any consistent metaphor. Is `relay` using the client/server metaphor? Are `replace` and `introduce` using a data structure metaphor (specifically, a stack)? `transform` is a bit better, but it sure has a hell of a type for something with that simple a name!

`stateful` is better, since it’s a name with an immediately obvious meaning, but I don’t actually understand why it needs to exist at all. It could be done just as easily using `reinterpret` combined with `runState`. It seems to run counter to the idea that most people are going to be defining effects in terms of existing ones, anyway. (The name is also not a verb.)

To summarize a bit, and to put things more bluntly, I don’t like these name changes, but I’m also not dead set against them. Obviously, you found the original names less helpful than I intended, so just because *I* think they follow a cohesive metaphor doesn’t they do. On the other hand, I’ve gotten a lot of very positive feedback from users of `freer-simple`’s API who have said the metaphor has been helpful to them, so I’d rather not throw it all out without some better understanding about what your philosophy is behind your proposed API.

File: src/Control/Monad/Freer/NonDet.hs (hunk 6 above)
Comment (lexi-lambda) [reply src/Control/Monad/Freer/NonDet.hs]: No idea, I’ve never personally used the `NonDet` effect in any of my projects. To be honest, I don’t even know what this function is supposed to do.

File: .travis.yml [hunk 15]
```diff
@@ -3,6 +3,9 @@ language: c
 cache:
   directories:
   - $HOME/.stack
+  - $HOME/.ghc
+  - $HOME/.cabal
+  - .stack-work
```
[the remark this answers is not in the document]
Comment (isovector) [reply .travis.yml]: I just cargo cult my CI, but it seems weird that every one of your builds takes 2.5h+.

File: src/Control/Monad/Freer.hs (hunk 1 above)
Comment (isovector) [reply src/Control/Monad/Freer.hs]: "Most people" is poorly phrased. I meant "most modules" want only to use effects, many fewer want to interpret them.

File: src/Control/Monad/Freer/Error.hs (hunk 2 above)
Comment (isovector) [reply src/Control/Monad/Freer/Error.hs]: Just that I didn't implement it.

File: src/Control/Monad/Freer/Interpretation.hs (hunk 5 above)
Comment (isovector) [reply src/Control/Monad/Freer/Interpretation.hs]: Thanks for your thoughts! TBH I implemented these in a greenfield, and so they're not renamed so much as "the first things I came up with." Your point about `stateful` is well taken---mine performs better than the two passes, but this could just as easily be a rewrite rule.

Comment (isovector) [reply src/Control/Monad/Freer/Interpretation.hs]: I don't have a particularly strong horse in this race, but would like to respond to a few points:

While a metaphor might be useful, it requires me to buy into it. Something I hate about `pipes` and `conduits` eg. is that in order to use them I need to read a really big tutorial to understand what the heck they're trying to talk about. In comparison, `streaming` doesn't require a metaphor, it's just "here are things you already know, but they're streaming."

Given the `streaming` point of view, `natural` is a good name _because_ it accurately describes what it does, assuming I already know what a natural transformation is. `transform` likewise lifts the effects of a monad transformer directly into an interpretation.

While I don't often agree with Ed Kmett's documentation strategy, one thing I do like is his inclination to "call something by its name." Given that natural transformations and monad transformers are particularly useful ideas _outside_ of this domain, maintaining that clarity seems valuable to me.

**Comment by isovector** [comment]
One idea I'm toying strongly with is to release the core `Eff` monad and interpretations in its own package, and then change this PR to add a dependency and implement your combinators in terms of those. 

The pros:

* We maintain 100% compatibility between the two projects
* I get my wish to separate the core machinery from standard effects, and thus not privilege any effects
* Neither of us needs to concede on naming things :)

This seems roughly analogous with the streaming library situation; they're all a little different in terms of guiding principles, but interop is just a function call away.

Thoughts?

File: src/Control/Monad/Freer/Interpretation.hs (hunk 5 above)
Comment (lexi-lambda) [reply src/Control/Monad/Freer/Interpretation.hs]: I understand. I think it’s probably more or less impossible to know if the metaphor is actually a net positive or net negative without doing some kind of rigorous study involving actual people, and that’s obviously not going to happen anytime soon, so we’re stuck with anecdotal evidence and personal feelings. I think the point of view you give is quite reasonable and has merits; I even agree with it a lot of the time.

But: given I don’t know whether the existing nomenclature is good or not, and I don’t think either of us can provide much evidence that something different would be meaningfully superior to the “newbie Haskeller” (since obviously neither of us are that), my preference is to avoid changing things up. I’d like to stick with the names we have now.

File: .travis.yml (hunk 15 above)
[the remark this answers is not in the document]
Comment (lexi-lambda) [reply .travis.yml]: I think the reason the builds take so long is mostly just cache misses. Travis has some rules around whether or not it will use a cache at all, based on some fairly conservative rules that I don’t totally understand. Builds that do use the cache are pretty fast, though, such as [this one](https://travis-ci.org/lexi-lambda/freer-simple/builds/445210407), so I don’t think caching more will help much here.

**Comment by lexi-lambda** [comment]
Sorry for taking so long to get to this. I’ve had some things going on, but I’m hoping I’ll be able to be significantly more responsive going forward.

> One idea I'm toying strongly with is to release the core Eff monad and interpretations in its own package, and then change this PR to add a dependency and implement your combinators in terms of those.

I have mixed feelings about this. In some respect, you’re probably right that there are some nontrivial merits to what you’re describing. On the other hand, I wish we could work through the social problem and agree to standardize around something in the effects space, and your proposal specifically makes it easier to *not* do that. I know that in this case, it doesn’t seem like a big deal, and maybe it even seems advantageous—we solve the technical problem of lack of interoperation between different parts of the effect ecosystem—but I worry about the social fragmentation it might lead to by discouraging the development of a shared vocabulary for these concepts.

It would be one thing if we had two competing ecosystems, both of which have seen large adoption, like `pipes` and `conduit`. In a case like that, I’d say: absolutely, if it’s possible to extract a shared core implementation that allows interoperation between those two ecosystems, that’d be a huge win. But we don’t really have that in the extensible effects world, and this seems too premature to me. I don’t get the sense we’d see any existing effects packages switching to the shared backend, so it’d just fragment the ecosystem further with even *more* frontend APIs.

Plus, the flexibility of the package being self-contained is nice. It means making backwards-incompatible changes, even to the core, doesn’t involve a very complicated process of synchronizing version numbers across projects and ensuring that everything stays buildable in all configurations. Instead, we can just make the change, bump one of the major version numbers, and cut a release. I think that’s a win.

**Comment by isovector** [comment]
Fair enough! I've spent the last few weeks thinking long and hard about effect systems. Since then I've come around to viewing your `Control.Monad.Freer` combinators as better fundamental choices than the `freer-effects` ones. Yours offer better composition and adequately solve my "I wish these things would compose better" complaint.

**Comment by isovector** [comment]
Some additional thoughts:

At the same time, I've been working my way through the (far too complicated) internals of `fused-effects`. I like that they're capable of providing second-order effects, but the conceptual cost of implementation is staggeringly high. 

It seems to me like we could lift second-order effects into `freer-simple` to give advanced-users the technology, but focus most of the technology on first-order effects. The `Lift` combinator from `fused-effects` does this transformation from first-to-second order.

That being said, their `bracket` implementation is still pretty wonky, which to me is the primary use-case. Their `bracket` will successfully trap short-circuiting effects, but incorrectly thread state. There are probably bespoke solutions for individual use-cases, but I'm not convinced the effect is doable in general (short of some new primitives to explicit set the IO masking state).

**Comment by isovector** [comment]
All of this is to say---how willing are you to introduce higher-order effects to `freer-simple` if we can do it without affecting the conceptual complexity for first-order users? How willing are you, even if it means we *still* can't write a satisfactory `bracket`?

**Comment by lexi-lambda** [comment]
> Since then I've come around to viewing your `Control.Monad.Freer` combinators as better fundamental choices than the `freer-effects` ones. Yours offer better composition and adequately solve my "I wish these things would compose better" complaint.

Hooray! :)

> At the same time, I've been working my way through the (far too complicated) internals of `fused-effects`. I like that they're capable of providing second-order effects, but the conceptual cost of implementation is staggeringly high.

This has largely been my impression, too, though I haven’t taken the time to examine `fused-effects` very closely, so I don’t really understand all the details. It seems really neat, but it does seem to prioritize screaming fast performance over interface simplicity… which is quite the opposite goal of this package!

That said, while I haven’t ever had the opportunity to *use* higher-order effects, and certainly not in anger, they seem like a strict improvement with no inherent downsides. If you could devise a way to make them work without making `freer-simple`’s existing API more complicated, that’d be awesome. It seems like, even if that doesn’t let us have `bracket`, it’d still be a net win, no?

**Comment by isovector** [comment]
Just kidding---`bracket` in `fused-effects` works fine, I was just doing it wrong.

I'm reasonably sure I can get the `freer-simple` API working with the performance and capabilities of `fused-effects`. Let's hold off on this PR until I have something to back up that claim :)

**Comment by lexi-lambda** [comment]
Okay—that sounds exciting! Should I hold my breath? :)
